### PR TITLE
Taking shape

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Introduce a new no_std SAX-style push parser along with extensive conformance, stress, and edge-case tests, and enhance error handling in the core library.

New Features:
- Add a SAX-style PushParser implementation with handler-based API and associated PushParseError type
- Export PushParser, PushParseError, PushParserHandler, ChunkReader and DefaultConfig in the public API

Enhancements:
- Extend ParseError with Utf8 and InputBufferFull variants and add From<ujson::Error> conversion

Tests:
- Add comprehensive conformance tests for PushParser and StreamParser against the JSON Checker suite
- Add stress tests for PushParser covering various buffer sizes, chunk patterns, and unicode edge cases
- Add minimal reproduction tests for InvalidSliceBounds issue
- Add copy-on-escape optimization tests for borrowed vs unescaped string lifetimes
- Add dedicated tests for escape sequence and unicode processing in PushParser

## Summary by CodeRabbit

* **New Features**
  * Introduced a new incremental JSON push parser with event-driven parsing and efficient memory management.
  * Added enhanced error types for UTF-8 errors and input buffer capacity issues.
  * Extended the public API with new parser interfaces and handler traits for customizable event processing.

* **Bug Fixes**
  * Improved robustness in handling buffer boundaries and Unicode escape sequences during JSON parsing.

* **Tests**
  * Added comprehensive conformance, unit, edge case, and stress tests for the push parser, including escape sequence handling, buffer isolation, and Unicode processing.
  * Developed new test utilities and handlers to support thorough parser validation across diverse scenarios.

### CODE REVIEW

File: picojson/src/push_parser.rs [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+            escape_state: EscapeState::None,
+            position_offset: 0,
+            current_position: 0,
+            token_start_pos: 0,
+            using_unescaped_buffer: false,
+        }
+    }
+
+    /// Processes a chunk of input data.
+    pub fn write<'input, E>(&mut self, data: &'input [u8]) -> Result<(), PushParseError<E>>
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs:72]: No, the tokenizer only ever emits at most 2 events per input byte
